File: android-native/app/src/main/java/com/example/kpkn/services/workout/WorkoutRestAlertRules.kt
package com.example.kpkn.services.workout

import com.example.kpkn.data.models.HapticIntensity

internal const val WORKOUT_REST_PREALERT_LEAD_SECONDS = 3
internal const val WORKOUT_REST_AUDIO_FAILURE_WINDOW_MS = 15 * 60 * 1000L

internal data class WorkoutRestToneStep(
    val durationMs: Int,
    val frequencyHz: Double,
    val volume: Double = 0.62,
    val pauseAfterMs: Long = 0L,
)

internal fun shouldScheduleWorkoutRestPrealert(
    durationSeconds: Int,
    leadSeconds: Int = WORKOUT_REST_PREALERT_LEAD_SECONDS,
): Boolean = durationSeconds > leadSeconds

internal fun workoutRestPrealertTriggerAt(
    endAtMs: Long,
    leadSeconds: Int = WORKOUT_REST_PREALERT_LEAD_SECONDS,
): Long = endAtMs - (leadSeconds * 1000L)

internal fun isRecentWorkoutRestAudioFailure(
    lastFailureAtMs: Long,
    nowMs: Long = System.currentTimeMillis(),
    windowMs: Long = WORKOUT_REST_AUDIO_FAILURE_WINDOW_MS,
): Boolean {
    if (lastFailureAtMs <= 0L) return false
    return nowMs - lastFailureAtMs in 0L..windowMs
}

internal fun workoutPrealertTonePlan(): List<WorkoutRestToneStep> = listOf(
    WorkoutRestToneStep(durationMs = 120, frequencyHz = 587.33, volume = 0.60, pauseAfterMs = 880L), // 3s
    WorkoutRestToneStep(durationMs = 120, frequencyHz = 587.33, volume = 0.60, pauseAfterMs = 880L), // 2s
    WorkoutRestToneStep(durationMs = 120, frequencyHz = 587.33, volume = 0.60, pauseAfterMs = 0L),   // 1s
)

internal fun workoutCompletionTonePlan(): List<WorkoutRestToneStep> = listOf(
    WorkoutRestToneStep(durationMs = 140, frequencyHz = 587.33, volume = 0.70, pauseAfterMs = 60L), // D5
    WorkoutRestToneStep(durationMs = 160, frequencyHz = 698.46, volume = 0.80, pauseAfterMs = 60L), // F5
    WorkoutRestToneStep(durationMs = 300, frequencyHz = 880.00, volume = 0.95, pauseAfterMs = 0L),  // A5
)

internal fun workoutPrealertVibrationPattern(): LongArray = longArrayOf(
    0L, 100L, // 3s
    900L, 100L, // 2s
    900L, 100L // 1s
)

internal fun workoutCompletionVibrationPattern(intensity: HapticIntensity): LongArray = when (intensity) {
    HapticIntensity.LIGHT -> longArrayOf(0L, 200L, 100L, 200L, 100L, 200L)
    HapticIntensity.MEDIUM -> longArrayOf(0L, 350L, 150L, 350L, 150L, 350L)
    HapticIntensity.STRONG -> longArrayOf(0L, 500L, 150L, 500L, 150L, 800L)
}

internal fun workoutCompletionVibrationAmplitudes(intensity: HapticIntensity): IntArray = when (intensity) {
    HapticIntensity.LIGHT -> intArrayOf(0, 180, 0, 180)
    HapticIntensity.MEDIUM -> intArrayOf(0, 220, 0, 220)
    HapticIntensity.STRONG -> intArrayOf(0, 255, 0, 255)
}

File: ios-native/KPKNFit/KPKNFit/Services/Workout/WorkoutRestAlertRules.swift
import Foundation

internal let WORKOUT_REST_PREALERT_LEAD_SECONDS = 3
internal let WORKOUT_REST_AUDIO_FAILURE_WINDOW_MS: Int64 = 15 * 60 * 1000

internal struct WorkoutRestToneStep {
    let durationMs: Int
    let frequencyHz: Double
    let volume: Double
    let pauseAfterMs: Int64

    init(durationMs: Int, frequencyHz: Double, volume: Double = 0.62, pauseAfterMs: Int64 = 0) {
        self.durationMs = durationMs
        self.frequencyHz = frequencyHz
        self.volume = volume
        self.pauseAfterMs = pauseAfterMs
    }
}

internal func shouldScheduleWorkoutRestPrealert(
    durationSeconds: Int,
    leadSeconds: Int = WORKOUT_REST_PREALERT_LEAD_SECONDS
) -> Bool {
    durationSeconds > leadSeconds
}

internal func workoutRestPrealertTriggerAt(
    endAtMs: Int64,
    leadSeconds: Int = WORKOUT_REST_PREALERT_LEAD_SECONDS
) -> Int64 {
    endAtMs - Int64(leadSeconds) * 1000
}

internal func isRecentWorkoutRestAudioFailure(
    lastFailureAtMs: Int64,
    nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
    windowMs: Int64 = WORKOUT_REST_AUDIO_FAILURE_WINDOW_MS
) -> Bool {
    guard lastFailureAtMs > 0 else { return false }
    let delta = nowMs - lastFailureAtMs
    return delta >= 0 && delta <= windowMs
}

internal func workoutPrealertTonePlan() -> [WorkoutRestToneStep] {
    [
        WorkoutRestToneStep(durationMs: 120, frequencyHz: 587.33, volume: 0.60, pauseAfterMs: 880),
        WorkoutRestToneStep(durationMs: 120, frequencyHz: 587.33, volume: 0.60, pauseAfterMs: 880),
        WorkoutRestToneStep(durationMs: 120, frequencyHz: 587.33, volume: 0.60, pauseAfterMs: 0),
    ]
}

internal func workoutCompletionTonePlan() -> [WorkoutRestToneStep] {
    [
        WorkoutRestToneStep(durationMs: 140, frequencyHz: 587.33, volume: 0.70, pauseAfterMs: 60),
        WorkoutRestToneStep(durationMs: 160, frequencyHz: 698.46, volume: 0.80, pauseAfterMs: 60),
        WorkoutRestToneStep(durationMs: 300, frequencyHz: 880.00, volume: 0.95, pauseAfterMs: 0),
    ]
}

internal func workoutPrealertVibrationPattern() -> [Int64] {
    [0, 100, 900, 100, 900, 100]
}

internal func workoutCompletionVibrationPattern(intensity: HapticIntensity) -> [Int64] {
    switch intensity {
    case .LIGHT:  return [0, 200, 100, 200, 100, 200]
    case .MEDIUM: return [0, 350, 150, 350, 150, 350]
    case .STRONG: return [0, 500, 150, 500, 150, 800]
    }
}

internal func workoutCompletionVibrationAmplitudes(intensity: HapticIntensity) -> [Int] {
    switch intensity {
    case .LIGHT:  return [0, 180, 0, 180]
    case .MEDIUM: return [0, 220, 0, 220]
    case .STRONG: return [0, 255, 0, 255]
    }
}
